### PR TITLE
Trackable completable futures

### PR DESCRIPTION
**Pull request by gregschohn**
### Description
Make it easier to debug why a CompletableFuture may be wedged.

### Issues Resolved
[List any issues this PR will resolve]

Is this a backport? If so, please add backport PR # and/or commits #

### Testing
[Please provide details of testing done: unit testing, integration testing and manual testing]

### Check List
- [ ] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/CapturedTrafficToHttpTransactionAccumulator.java [hunk 1]
```diff
@@ -177,34 +185,36 @@ private Accumulation getAccumulationForFirstRequestObservation(String nodeId,
...
-                accumulation.state = Accumulation.State.REQUEST_SENT;
-                return true;
-            case REQUEST_SENT:
-                fullDataHandler.accept(accumulation.rrPair);
-                accumulation.state = Accumulation.State.RESPONSE_SENT;
-                liveStreams.remove(nodeId, connectionId);
-        }
-        return false;
+    private void handleEndOfRequest(String connectionId, Accumulation accumulation) {
+        assert accumulation.state == Accumulation.State.NOTHING_SENT : "state == " + accumulation.state;
```
Comment (lewijacn) [inline TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/CapturedTrafficToHttpTransactionAccumulator.java:189]: What is actually happening here for my own knowledge? Or I guess more specifically what does a colon by itself do here?

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/TrafficReplayer.java [hunk 2]
```diff
@@ -164,30 +169,43 @@ private void runReplayWithIOStreams(Duration observedPacketConnectionTimeout,
...
+                                                        } finally {
+                                                            requestToFinalWorkFuturesMap.remove(rrPair.requestData);
+                                                            log.error("removed rrPair.requestData to requestToFinalWorkFuturesMap for " +
+                                                                    rrPair.connectionId);
+                                                        }
+                                                    }), ()->"TrafficReplayer.runReplayWithIOStreams.progressTracker");
+                            if (!resultantCf.future.isDone()) {
+                                log.error("Adding " + rrPair.connectionId + " to requestToFinalWorkFuturesMap");
+                                requestToFinalWorkFuturesMap.put(rrPair.requestData, resultantCf);
+                                if (resultantCf.future.isDone()) {
```
Comment (lewijacn) [inline TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/TrafficReplayer.java:205]: What is point of this if statement, to potentially catch futures that have completed since lines 203 and 204?

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/TrafficReplayer.java [hunk 3]
```diff
@@ -197,37 +215,92 @@ private void runReplayWithIOStreams(Duration observedPacketConnectionTimeout,
...
-                    .handle((t, v) -> {
-                        log.info("stopping packetHandlerFactory's group");
-                        packetHandlerFactory.stopGroup();
-                        return null; // squash exceptions for individual requests
-                    })
-                    .get();
+            var PRIMARY_LOG_LEVEL = Level.INFO;
+            var SECONDARY_LOG_LEVEL = Level.INFO;
+            var logLevel = PRIMARY_LOG_LEVEL;
+            for (var timeout = Duration.ofSeconds(1); ; timeout = timeout.multipliedBy(2)) {
```
Comment (lewijacn) [inline TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/TrafficReplayer.java:221]: This is hard to follow, not sure what we are looping here

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/CapturedTrafficToHttpTransactionAccumulator.java (hunk 1 above)
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/CapturedTrafficToHttpTransactionAccumulator.java:189]: The part after the colon is evaluated at runtime and is included in the Assertion failure message when the assertion (before the colon) fails.

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/TrafficReplayer.java (hunk 2 above)
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/TrafficReplayer.java:205]: Yep - to deal with race conditions.  I don't think that I have a unit test to deal with this, but I did notice that when dealing w/ my static input file, this happened to a number of entries that just lingered in and were wasting space in the allRemainingWork checks at the end (they were complete, so it wasn't a big deal, but it created more noise for when I was debugging)

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/TrafficReplayer.java (hunk 3 above)
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/TrafficReplayer.java:221]: We're running waitForRemainingWork repeatedly.  First we do it with a 1 second timeout, then we double it indefinitely.  We'll also change the log level that we want to report what's still remaining after the first run - the thought being that after the first second or whatever the wait duration was, "normal" things should be wrapped up pretty well.

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/datahandlers/http/NettySendByteBufsToPacketHandlerHandler.java [hunk 4]
```diff
@@ -25,61 +27,80 @@ public class NettySendByteBufsToPacketHandlerHandler extends ChannelInboundHandl
...
-        CompletableFuture<AggregatedTransformedResponse> packetReceiverCompletionFuture =
-                new CompletableFuture<>();
-        if (currentFuture.isDone()) {
-            if (currentFuture.isCompletedExceptionally()) {
-                currentFuture.handle((v,t)->
-                        packetReceiverCompletionFuture.completeExceptionally(t));
-                packetReceiverCompletionFutureRef.set(packetReceiverCompletionFuture);
+//                        () -> "Waiting for NettySendByteBufsToPacketHandlerHandler to complete this future.  " +
+//                                "this.currentFuture, whose completion will complete this future is currently:\n" +
+//                                "    <<" + currentFuture + ">>\n");
```
Comment (mikaylathompson) [inline TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/datahandlers/http/NettySendByteBufsToPacketHandlerHandler.java:49]: What is this related to? Intentional that it's here?

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/datahandlers/http/NettySendByteBufsToPacketHandlerHandler.java [hunk 5]
```diff
@@ -93,16 +114,25 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) {
                     " ctx hash=" + System.identityHashCode(ctx));
             var bb = ((ByteBuf) msg).retain();
             log.trace("CR: old currentFuture="+currentFuture);
+            // I don't want to capture this object, but the preexisting value of the currentFuture field
```
Comment (mikaylathompson) [inline TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/datahandlers/http/NettySendByteBufsToPacketHandlerHandler.java:117]: The second half of this comment seems to match up with the code, but what is "this object" in the first half referring to?

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/util/DiagnosticTrackableCompletableFuture.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    public String toString() {
+        return formatAsString(x->null);
+    }
+
+    public String formatAsString(@NonNull Function<DiagnosticTrackableCompletableFuture,String> resultFormatter) {
+        var strList = diagnosticStream().map(kvp->formatFutureWithDiagnostics(kvp, resultFormatter))
+                .collect(Collectors.toList());
+        return strList.stream().collect(Collectors.joining("<-"));
+    }
+//"[…]"
```
Comment (mikaylathompson) [inline TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/util/DiagnosticTrackableCompletableFuture.java:175]: accidental or a weird Java thing I don't know?

**Review by mikaylathompson — APPROVED** [review]
I think I ~80% understand what's going on in all of this. I have a couple comments that lean towards increasing my understanding of what's going on. I don't think any are blockers.

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/TrafficReplayer.java (hunk 3 above)
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/TrafficReplayer.java:221]: Ahhh - gotcha.. You're right.  Thanks.

File: TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/TrafficReplayer.java [hunk 7]
```diff
@@ -164,30 +169,43 @@ private void runReplayWithIOStreams(Duration observedPacketConnectionTimeout,
...
+                                                            AggregatedTransformedResponse rval =
+                                                                    packageAndWriteResponse(tupleWriter, rrPair, summary, t);
+                                                            successCount.incrementAndGet();
+                                                            return rval;
+                                                        } catch (Exception e) {
+                                                            exceptionCount.incrementAndGet();
+                                                            throw e;
+                                                        } finally {
+                                                            requestToFinalWorkFuturesMap.remove(rrPair.requestData);
+                                                            log.error("removed rrPair.requestData to requestToFinalWorkFuturesMap for " +
```
[the remark this answers is not in the document]
Comment (gregschohn) [reply TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/TrafficReplayer.java:198]: Agree - definitely didn't mean to leave it at error!  Reduced to trace
